## Project 2

### Tidying Data

#### Project Task:

(1) Choose any three of the “wide” datasets identified in the Week 5 Discussion items. (You may use your own dataset; please don’t use my Sample Post dataset, since that was used in your Project 1 assignment!) For each of the three chosen datasets:
* Create a .CSV file (or optionally, a MySQL database!) that includes all of the information included in the dataset. You’re encouraged to use a “wide” structure similar to how the information appears in the discussion item, so that you can practice tidying and transformations as described below.
* Read the information from your .CSV file into Python, and use pandas as needed to tidy and transform your data. [Most of your grade will be based on this step!]
* Perform the analysis requested in the discussion item.
* Your code should be in an Jupyter Notebook, posted to your GitHub repository, and should include narrative descriptions of your data cleanup work, analysis, and conclusions.

(2) Please include the URL for the Jupyter Notebook in your homework submission, for each of the three chosen datasets. You may create a single Jupyter Notebook, or three separate notebooks.

In [1]:
import numpy as np
import pandas as pd
import re

##### Dataset #1:

Leading Causes of Death in New York City

- Which cause of death had most deaths in 2014?
- Which cause of death had the lowest death rate in 2014?
- What are the top 5 causes of death for Asian and Pacific Islander in 2014?

In [2]:
deaths = pd.read_csv(r'C:\Users\DJEli112\Desktop\CUNY SPS\Spring 2017\IS 362\New_York_City_Leading_Causes_of_Death.csv')

After importing the CSV file, I wanted to see the beginnng 5 rows of this table by utilizing .head():

In [3]:
deaths.head()

Year                              Leading Cause Sex         Race Ethnicity  \
0  2014                Diabetes Mellitus (E10-E14)   F  Other Race/ Ethnicity   
1  2011  Cerebrovascular Disease (Stroke: I60-I69)   M     White Non-Hispanic   
2  2008      Malignant Neoplasms (Cancer: C00-C97)   M     Not Stated/Unknown   
3  2010      Malignant Neoplasms (Cancer: C00-C97)   F               Hispanic   
4  2012  Cerebrovascular Disease (Stroke: I60-I69)   M     Black Non-Hispanic   

  Deaths Death Rate Age Adjusted Death Rate  
0     11          .                       .  
1    290       21.7                    18.2  
2     60          .                       .  
3   1045       85.9                    98.5  
4    170       19.9                    23.3

As you can tell, there are (.) dots or periods that are in place of a null value that is needed. Since there is a dot, I believe it is important to check the data types for each column:

In [4]:
deaths.dtypes

Year                        int64
Leading Cause              object
Sex                        object
Race Ethnicity             object
Deaths                     object
Death Rate                 object
Age Adjusted Death Rate    object
dtype: object

Since each column has an object type for Deaths, Deaths, Rates, and Age Adjusted Death Rate, it is better to replace the dots with null values and then convert the rows for the respective columns into an integer or float through .to_numeric:

In [5]:
deaths = deaths.replace(to_replace='.', value=np.nan)

In [6]:
deaths[['Deaths', 'Death Rate', 'Age Adjusted Death Rate']] = deaths[['Deaths', 'Death Rate', 
                                                              'Age Adjusted Death Rate']].apply(pd.to_numeric)

deaths.dtypes

Year                         int64
Leading Cause               object
Sex                         object
Race Ethnicity              object
Deaths                     float64
Death Rate                 float64
Age Adjusted Death Rate    float64
dtype: object

#### Which cause of death had most deaths in 2014?

In [7]:
deaths[deaths['Year']==2014].sort_values('Deaths', ascending=False).head(1)

Year                                   Leading Cause Sex  \
620  2014  Diseases of Heart (I00-I09, I11, I13, I20-I51)   F   

         Race Ethnicity  Deaths  Death Rate  Age Adjusted Death Rate  
620  White Non-Hispanic  4507.0       318.0                    161.0

After converting the dots to the null values, I successfully determined that heart disease had the most deaths at 4507 in 2014. I did this by sorting for Deaths, where the year is 2014, and the top count of deaths by setting ascending to False and .head(1).

#### Which cause of death had the lowest death rate in 2014?

In [8]:
deaths[deaths['Year']==2014].sort_values('Death Rate').head(1)

Year              Leading Cause Sex              Race Ethnicity  Deaths  \
342  2014  Alzheimer's Disease (G30)   F  Asian and Pacific Islander    26.0   

     Death Rate  Age Adjusted Death Rate  
342         4.2                      4.7

In this instance, I sorted by Death Rate in which the year is 2014 with the bottom count of death rates by either setting ascending to True or leaving it blank, as well as .head(1). The result shows Alzheimer's Disease at 4.2.

#### What are the top 5 causes of death for Asian and Pacific Islander in 2014?

In [9]:
deaths[(deaths['Year']==2014) & (deaths['Race Ethnicity']=='Asian and Pacific Islander')].sort_values('Deaths', 
                                                                                          ascending=False).head(5)

Year                                   Leading Cause Sex  \
210  2014           Malignant Neoplasms (Cancer: C00-C97)   M   
988  2014  Diseases of Heart (I00-I09, I11, I13, I20-I51)   M   
344  2014           Malignant Neoplasms (Cancer: C00-C97)   F   
567  2014  Diseases of Heart (I00-I09, I11, I13, I20-I51)   F   
54   2014                                All Other Causes   M   

                 Race Ethnicity  Deaths  Death Rate  Age Adjusted Death Rate  
210  Asian and Pacific Islander   657.0       114.5                    129.5  
988  Asian and Pacific Islander   554.0        96.5                    118.5  
344  Asian and Pacific Islander   502.0        80.2                     80.6  
567  Asian and Pacific Islander   462.0        73.8                     81.1  
54   Asian and Pacific Islander   424.0        73.9                     90.4

Since there are multiple parts to this question, I needed to decide how to combine how to sort by deaths and define a specific ethnicity at a certain year for the top 5 causes of death. I sorted by Deaths in which the year is 2014 and the Race Ethnicity is Asian and Pacific Islander with the top 5 count of Deaths by setting ascending at False and .head(5). As a result, Malignant Neoplasms, Diseases of Heart, Malignant Neoplasms, Diseases of Heart and All Other Causes have respectives deaths of 657, 554, 502, 462, and 424.

##### Dataset # 2:

California Cities

- Which cities have the greatest amount of open water?
- Which cities have the highest elevation?
- Which cities have the greatest ratio of land area to water?

In [10]:
cities = pd.read_csv(r'C:\Users\DJEli112\Desktop\CUNY SPS\Spring 2017\IS 362\california_cities.csv')

As with the previous data set, I will check the beginning 5 rows of the California Cities data set by utilizing .head():

In [11]:
cities.head()

Unnamed: 0         city       latd       longd  elevation_m  elevation_ft  \
0           0     Adelanto  34.576111 -117.432778        875.0        2871.0   
1           1  AgouraHills  34.153333 -118.761667        281.0         922.0   
2           2      Alameda  37.756111 -122.274444          NaN          33.0   
3           3       Albany  37.886944 -122.297778          NaN          43.0   
4           4     Alhambra  34.081944 -118.135000        150.0         492.0   

   population_total  area_total_sq_mi  area_land_sq_mi  area_water_sq_mi  \
0             31765            56.027           56.009             0.018   
1             20330             7.822            7.793             0.029   
2             75467            22.960           10.611            12.349   
3             18969             5.465            1.788             3.677   
4             83089             7.632            7.631             0.001   

   area_total_km2  area_land_km2  area_water_km2  area_water_percent  
0         145.107        145.062           0.046                0.03  
1          20.260         20.184           0.076                0.37  
2          59.465         27.482          31.983               53.79  
3          14.155          4.632           9.524               67.28  
4          19.766         19.763           0.003                0.01

As you can tell, there is an unnammed column with rows that are not pertinent to the dataset itself. Another observation is AgouraHills not seperated by a space. Unlike the previous table, the null values that are inserted in the table and does not need to be converted. In order to tidy up the unnammed column, rename Unnamed: 0 to unnamed by renaming it and dropping the column. To seperate a space in between AgouraHills and other cities, a function will defined to seperate a space for all cities in California for this data set.

In [12]:
cities_cols =['unnamed', 'city', 'latd', 'longd', 'elevation_m', 'elevation_ft',
       'population_total', 'area_total_sq_mi', 'area_land_sq_mi',
       'area_water_sq_mi', 'area_total_km2', 'area_land_km2', 'area_water_km2',
       'area_water_percent']
cities.columns = cities_cols
cities.head()

unnamed         city       latd       longd  elevation_m  elevation_ft  \
0        0     Adelanto  34.576111 -117.432778        875.0        2871.0   
1        1  AgouraHills  34.153333 -118.761667        281.0         922.0   
2        2      Alameda  37.756111 -122.274444          NaN          33.0   
3        3       Albany  37.886944 -122.297778          NaN          43.0   
4        4     Alhambra  34.081944 -118.135000        150.0         492.0   

   population_total  area_total_sq_mi  area_land_sq_mi  area_water_sq_mi  \
0             31765            56.027           56.009             0.018   
1             20330             7.822            7.793             0.029   
2             75467            22.960           10.611            12.349   
3             18969             5.465            1.788             3.677   
4             83089             7.632            7.631             0.001   

   area_total_km2  area_land_km2  area_water_km2  area_water_percent  
0         145.107        145.062           0.046                0.03  
1          20.260         20.184           0.076                0.37  
2          59.465         27.482          31.983               53.79  
3          14.155          4.632           9.524               67.28  
4          19.766         19.763           0.003                0.01

As previously stated, the "Unnamed: 0" column has been successfully renamed to "unnamed". In order to drop this column, utilize .drop() with specific parameters and show the first 5 rows with .head():

In [13]:
cities.drop('unnamed', axis=1, inplace=True)
cities.head()

city       latd       longd  elevation_m  elevation_ft  \
0     Adelanto  34.576111 -117.432778        875.0        2871.0   
1  AgouraHills  34.153333 -118.761667        281.0         922.0   
2      Alameda  37.756111 -122.274444          NaN          33.0   
3       Albany  37.886944 -122.297778          NaN          43.0   
4     Alhambra  34.081944 -118.135000        150.0         492.0   

   population_total  area_total_sq_mi  area_land_sq_mi  area_water_sq_mi  \
0             31765            56.027           56.009             0.018   
1             20330             7.822            7.793             0.029   
2             75467            22.960           10.611            12.349   
3             18969             5.465            1.788             3.677   
4             83089             7.632            7.631             0.001   

   area_total_km2  area_land_km2  area_water_km2  area_water_percent  
0         145.107        145.062           0.046                0.03  
1          20.260         20.184           0.076                0.37  
2          59.465         27.482          31.983               53.79  
3          14.155          4.632           9.524               67.28  
4          19.766         19.763           0.003                0.01

As shown in the table, the "unnamed" column was dropped and looks more neater than before.

In [14]:
def city_spacing(city):
    spacing = re.findall('[A-Z][^A-Z]*', city)
    return ' '.join(spacing)

cities['city'] = cities['city'].apply(city_spacing)

cities['city'].head(5)

0        Adelanto
1    Agoura Hills
2         Alameda
3          Albany
4        Alhambra
Name: city, dtype: object

After importing re or regular expression, it is easier to work with strings and manipulate them to your liking, In the case, I wanted to add a space between two or more words for a city's name. Although this was applied to every city, the first 5 were shown, and AgouraHills is now "Agoura Hills."

#### Which cities have the greatest amount of open water?

In [15]:
cities[['city', 'area_water_sq_mi']].sort_values('area_water_sq_mi', ascending=False).head(5)

city  area_water_sq_mi
370  San Francisco           185.020
367      San Diego            47.210
239    Los Angeles            34.000
285  Newport Beach            29.173
93        Coronado            24.735

Since the question does not specify the number of cities, I will do the top 5 cities for each question. This shows San Francisco, San Diego, Los Angeles, Newport Beach, and Coronado as the cities with largest amount of open water that were sorted by area_water_sq_mi and set ascending to False.

#### Which cities have the highest elevation?

In [16]:
cities[['city', 'elevation_ft']].sort_values('elevation_ft', ascending=False).head(5)

city  elevation_ft
246     Mammoth Lakes        7880.0
39      Big Bear Lake        6752.0
416  South Lake Tahoe        6237.0
436           Truckee        5817.0
242          Loyalton        4951.0

In this instance, I sorted by elevation_ft and set ascending to False with the top 5 values shown for each respective city: Mammoth Lakes, Big Bear Lake, South Lake Tahoe, Truckee, and Loyalton.

#### Which cities have the greatest ratio of land area to water?

In [17]:
cities[['city', 'area_water_percent']].sort_values('area_water_percent', ascending=False).head(5)

city  area_water_percent
48        Brisbane               84.58
146    Foster City               81.07
366      Sand City               80.77
370  San Francisco               80.00
35       Belvedere               78.42

For this question, I sorted by area_water_percent and set acending to False with the top 5 values shown for each respective city: Brisbane, Foster City, Sand City, San Francisco, and Belvedere.

##### Dataset 3:

Planes

- Which plane manufacturer has the most seats?

In [18]:
planes = pd.read_csv(r'C:\Users\DJEli112\Desktop\CUNY SPS\Spring 2017\IS 362\planes.csv')

As with every dataset thus far, I displayed the top 5 rows by utilizing .head():

In [19]:
planes.head()

tailnum    year                     type      manufacturer      model  \
0  N10156  2004.0  Fixed wing multi engine           EMBRAER  EMB-145XR   
1  N102UW  1998.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
2  N103US  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
3  N104UW  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
4  N10575  2002.0  Fixed wing multi engine           EMBRAER  EMB-145LR   

   engines  seats  speed     engine  
0        2     55    NaN  Turbo-fan  
1        2    182    NaN  Turbo-fan  
2        2    182    NaN  Turbo-fan  
3        2    182    NaN  Turbo-fan  
4        2     55    NaN  Turbo-fan

Since the discussion item said to find the speed at which each plane moves, I researched the CSV and realized none of the planes had a speed and were associated with a null value. Instead, I will tidy the dataset by looking at the number of seats for each manufacturer. As shown above, the year has a float value with a decimal float that should be removed. This will be accomplished by applying a split of the decimal float and ultimately removing it for all the year values in the data set:

In [20]:
planes['year'] = planes['year'].astype(str).apply(lambda x: x.split('.')[0])

planes['year'] = pd.to_numeric(planes['year'], errors='ignore')

planes['year'].head(5)

0    2004
1    1998
2    1999
3    1999
4    2002
Name: year, dtype: object

Although it shows only the top five values, all of the year values have the decimal floats removed. Since the speeds for each planes are mostly null values and there are a lot of values to calculate, we will take a look at the seats grouped by manufacturer:

In [21]:
planes.groupby('manufacturer')['seats'].max()

manufacturer
AGUSTA SPA                         8
AIRBUS                           379
AIRBUS INDUSTRIE                 379
AMERICAN AIRCRAFT INC              2
AVIAT AIRCRAFT INC                 2
AVIONS MARCEL DASSAULT            12
BARKER JACK L                      2
BEECH                             10
BELL                              11
BOEING                           450
BOMBARDIER INC                    95
CANADAIR                          55
CANADAIR LTD                       2
CESSNA                             8
CIRRUS DESIGN CORP                 4
DEHAVILLAND                       16
DOUGLAS                          102
EMBRAER                           55
FRIEDEMANN JON                     2
GULFSTREAM AEROSPACE              22
HURLEY JAMES LARRY                 2
JOHN G HESS                        2
KILDALL GARY                       2
LAMBERT RICHARD                    2
LEARJET INC                       11
LEBLANC GLENN T                    2
MARZ BARRY               

In this output, we are shown every manufacturer and their respective seat numbers that are ordered alphabetically. To see how many planes for each manufacturer, we will show the counts for each manufacturer:

In [22]:
planes['manufacturer'].value_counts()

BOEING                           1630
AIRBUS INDUSTRIE                  400
BOMBARDIER INC                    368
AIRBUS                            336
EMBRAER                           299
MCDONNELL DOUGLAS                 120
MCDONNELL DOUGLAS AIRCRAFT CO     103
MCDONNELL DOUGLAS CORPORATION      14
CESSNA                              9
CANADAIR                            9
PIPER                               5
GULFSTREAM AEROSPACE                2
BEECH                               2
AMERICAN AIRCRAFT INC               2
BELL                                2
STEWART MACO                        2
HURLEY JAMES LARRY                  1
JOHN G HESS                         1
FRIEDEMANN JON                      1
MARZ BARRY                          1
CIRRUS DESIGN CORP                  1
BARKER JACK L                       1
AVIAT AIRCRAFT INC                  1
SIKORSKY                            1
PAIR MIKE E                         1
CANADAIR LTD                        1
LEARJET INC 

From this output, we can infer that Boeing will have the most seats since it has the most number of planes for this manufacturer. However, we can conclude this result by finding the sum of seats for each manufacturer:

In [23]:
sum_of_seats = planes.groupby('manufacturer')['seats'].sum()
sum_of_seats

manufacturer
AGUSTA SPA                            8
AIRBUS                            74324
AIRBUS INDUSTRIE                  74961
AMERICAN AIRCRAFT INC                 4
AVIAT AIRCRAFT INC                    2
AVIONS MARCEL DASSAULT               12
BARKER JACK L                         2
BEECH                                19
BELL                                 16
BOEING                           285556
BOMBARDIER INC                    27235
CANADAIR                            495
CANADAIR LTD                          2
CESSNA                               48
CIRRUS DESIGN CORP                    4
DEHAVILLAND                          16
DOUGLAS                             102
EMBRAER                           13645
FRIEDEMANN JON                        2
GULFSTREAM AEROSPACE                 44
HURLEY JAMES LARRY                    2
JOHN G HESS                           2
KILDALL GARY                          2
LAMBERT RICHARD                       2
LEARJET INC                

As shown above, the sum of seats for each manufacturer is sorted alphabetically. In order to analyze this further, we will sort the data to show the top 5 manufacturers with the most seats by setting ascending to False and .head(5):

In [24]:
sum_of_seats.sort_values(ascending=False).head(5)

manufacturer
BOEING               285556
AIRBUS INDUSTRIE      74961
AIRBUS                74324
BOMBARDIER INC        27235
MCDONNELL DOUGLAS     19446
Name: seats, dtype: int64

The data is now sorted by most seats for each of the 5 manufacturers: Boeing, Airbus Industrie, Airbus, Bombardier Inc, and McDonnell Douglas. To find the top plane manufacturer with the most seats, replaces .head(5) to .head(1):

In [25]:
sum_of_seats.sort_values(ascending=False).head(1)

manufacturer
BOEING    285556
Name: seats, dtype: int64

As confirmed, Boeing is the manufacturer that provided planes with the most seats because Boeing was the manufacturer with the most planes. 

### Final Analysis:

Although each dataset had its own method of tidying, I believe these skills are pertinent in data analysis and data science. The planes dataset was the only table in which it was difficult to figure out every speed for each plane since there are many values. Instead, I tried to find the plane manufacturer that had the most seats (Boeing). Pertaining to california cities, there were a few minor adjustments to display the table in a tidier manner such as spacing out the city names and dropping an unnamed, extra column. As for leading deaths in New York City, it was harder to determine how to replace the periods or dots with null values. After analzying the datatypes, the object datatype should be converted to a integer or float value as well as inserting null values for them. Overall, the project was an interesting approach in tidying several wide data sets.